src/test/regress/sql/failure_multi_dml.sql

```diff
@@ -0,0 +1,148 @@
+SELECT citus.mitmproxy('conn.allow()');
+
+SET citus.shard_count = 2;
+SET citus.shard_replication_factor = 1; -- one shard per worker
+SET citus.next_shard_id TO 103400;
+ALTER SEQUENCE pg_catalog.pg_dist_placement_placementid_seq RESTART 100;
+
+CREATE TABLE dml_test (id integer, name text);
+SELECT create_distributed_table('dml_test', 'id');
+
+COPY dml_test FROM STDIN WITH CSV;
+1,Alpha
+2,Beta
+3,Gamma
+4,Delta
+\.
+
+SELECT citus.clear_network_traffic();
+
+---- test multiple statements spanning multiple shards, at each possible point
+
+-- fail at DELETE
+SELECT citus.mitmproxy('conn.onQuery(query="^DELETE").kill()');
+
+BEGIN;
+DELETE FROM dml_test WHERE id = 1;
+DELETE FROM dml_test WHERE id = 2;
+INSERT INTO dml_test VALUES (5, 'Epsilon');
+UPDATE dml_test SET name = 'alpha' WHERE id = 1;
+UPDATE dml_test SET name = 'gamma' WHERE id = 3;
+COMMIT;
+
+--- shouldn't see any changes after failed DELETE
+SELECT * FROM dml_test ORDER BY id ASC;
+
+-- fail at INSERT
+SELECT citus.mitmproxy('conn.onQuery(query="^INSERT").kill()');
+
+BEGIN;
+DELETE FROM dml_test WHERE id = 1;
+DELETE FROM dml_test WHERE id = 2;
+INSERT INTO dml_test VALUES (5, 'Epsilon');
+UPDATE dml_test SET name = 'alpha' WHERE id = 1;
+UPDATE dml_test SET name = 'gamma' WHERE id = 3;
+COMMIT;
+
+--- shouldn't see any changes before failed INSERT
+SELECT * FROM dml_test ORDER BY id ASC;
+
+-- fail at UPDATE
+SELECT citus.mitmproxy('conn.onQuery(query="^UPDATE").kill()');
+
+BEGIN;
+DELETE FROM dml_test WHERE id = 1;
+DELETE FROM dml_test WHERE id = 2;
+INSERT INTO dml_test VALUES (5, 'Epsilon');
+UPDATE dml_test SET name = 'alpha' WHERE id = 1;
+UPDATE dml_test SET name = 'gamma' WHERE id = 3;
+COMMIT;
+
+--- shouldn't see any changes after failed UPDATE
+SELECT * FROM dml_test ORDER BY id ASC;
+
+-- FIXME: Can't test this because the prepared transaction ID has PID in it
```

Comment:
I might have an idea here, I will have a look tomorrow

Reply:
I have done some tests. I think the pid is in the name of the prepared transaction.

We can prevent that name to show up in the tests by a `SET LOCAL client_min_messages TO 'FATAL';` in the transaction block. This would prevent the error messages to show.

Even though the error message will not show in the log output we can still verify its behaviour by inspecting the contents of the table as we do with the other tests via the commented out `SELECT` statement.

Reply:
To give some more context on this, we had a similar discussion around testing this with @lithp  when he first implemented the infrastructure.

Basically, @lithp proposed to add a GUC, which is only set during failure tests running, to not add the PID in the 2PC identifiers. However, I pushed back because (a) that was the only source code change in Citus, and implementing the failure testing automation shouldn't have required any code changes (b) I didn't want to add any complexity to 2PC id generation codes just for testing purposes.

I proposed him another approach for writing those tests, you can see it [here](https://github.com/citusdata/citus/pull/2119#discussion_r190549470)

Reply:
Went with the `FATAL` approach for now.